webrender/res/ps_angle_gradient.fs.glsl

```diff
@@ -2,38 +2,17 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-float offset(int index) {
-    return vOffsets[index / 4][index % 4];
-}
-
-float linearStep(float lo, float hi, float x) {
-    float d = hi - lo;
-    float v = x - lo;
-    if (d != 0.0) {
-        v /= d;
-    }
-    return clamp(v, 0.0, 1.0);
-}
+uniform sampler2D sGradients;
 
 void main(void) {
-    float angle = atan(-vEndPoint.y + vStartPoint.y,
-                        vEndPoint.x - vStartPoint.x);
-    float sa = sin(angle);
-    float ca = cos(angle);
-
-    float sx = vStartPoint.x * ca - vStartPoint.y * sa;
-    float ex = vEndPoint.x * ca - vEndPoint.y * sa;
-    float d = ex - sx;
-
-    float x = vPos.x * ca - vPos.y * sa;
+    // Either saturate or modulo the offset depending on repeat mode, then scale to number of
+    // gradient color entries.
+    float x = mix(clamp(vOffset, 0.0, 1.0), fract(vOffset), vGradientRepeat) * vGradientScale;
 
-    oFragColor = mix(vColors[0],
-                     vColors[1],
-                     linearStep(sx + d * offset(0), sx + d * offset(1), x));
+    // Start at the center of first color in the nearest 2-color entry, then offset with the
+    // fractional remainder to interpolate between the colors.
+    x = 2.0 * floor(x) + 0.5 + fract(x);
```

Comment:
I could just remove that and base it on the textureSize since that will address your concern about gpu_data_count as well.